Sources/ServiceLifecycle/GracefulShutdown.swift

```diff
@@ -0,0 +1,116 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftServiceLifecycle open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftServiceLifecycle project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftServiceLifecycle project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/// Execute an operation with a graceful shutdown handler that’s immediately invoked if the current task is shutting down gracefully.
+///
+/// This doesn’t check for graceful shutdown, and always executes the passed operation.
+/// The operation executes on the calling execution context and does not suspend by itself, unless the code contained within the closure does.
+/// If graceful shutdown occurs while the operation is running, the graceful shutdown handler will execute concurrently with the operation.
+///
+/// When `withShutdownGracefulHandler` is used in a Task that has already been gracefully shutdown, the `onGracefulShutdown` handler
+/// will be executed immediately before operation gets to execute. This allows the `onGracefulShutdown` handler to set some external “shutdown” flag
+/// that the operation may be atomically checking for in order to avoid performing any actual work once the operation gets to run.
+///
+/// A common use-case is to listen to graceful shutdown and use the `ServerQuiescingHelper` from `swift-nio-extras` to
+/// trigger the quiescing sequence. Furthermore, graceful shutdown will propagate to any child task that is currently executing
+///
+/// - Parameters:
+///   - operation: The actual operation.
+///   - handler: The handler which is invoked once graceful shutdown has been triggered.
+@_unsafeInheritExecutor
```

Comment:
Do we really want that?

Reply:
Some offline discussion here:

I tried to use `@_inheritActorContext` first here but it didn't work. The magic piece is to also mark the closure as `@Sendable` at the same time. This solves the warnings; however, this makes the `withGracefulShutdown` method only work inside actors. Let's look at concrete code

```swift
final actor Foo {
    var bar = 1
    func run() async {
        await withShutdownGracefulHandler {
            bar += 1
        } onGracefulShutdown: {
        }
    }
}

final class FooBar {
    var bar = 1
    func run() async {
        await withShutdownGracefulHandler {
            self.bar += 1 // Produces a warning: Capture of 'self' with non-sendable type 'FooBar' in a `@Sendable` closure
        } onGracefulShutdown: {
        }
    }
}
```

The thing that we want to convey though is that the `operation` closure of the `withShutdownGracefulHandler` is running on the same executor as the calling context. That's why I reached for `@_unsafeInheritExecutor`; however, as it turns out that thing really is unsafe and doesn't hop at all anymore.

Conclusion:
For now the best thing to do is use the `@_inheritActorContext` and `@Sendable` on the `operation` closure which forces people to use this only in actors. I am also going to create an issue on the compiler that we need to be able to spell this kind of `with` method without getting warnings.